multinet/table.py

```diff
@@ -0,0 +1,109 @@
+"""Operations that deal with tables."""
+from __future__ import annotations  # noqa: T484
+from arango.collection import StandardCollection
+from arango.aql import AQL
+
+from multinet import util
+from multinet.types import EdgeTableProperties
+from multinet.errors import InternalServerError
+
+from typing import List, Set, Dict, Iterable, Union, Optional
+
+
+class Table:
+    """Table."""
+
+    def __init__(self, name: str, workspace: str, handle: StandardCollection, aql: AQL):
+        """Init Table class."""
+        self.name = name
+        self.workspace = workspace
+        self.handle = handle
+        self.aql = aql
+
+    def rows(self, offset: Optional[int] = None, limit: Optional[int] = None) -> Dict:
+        """Return the desired rows in a table."""
+        rows = self.handle.find({}, skip=offset, limit=limit)
+        count = self.handle.all().count()
+
+        return {"count": count, "rows": list(rows)}
+
+    def row(self, doc: Union[Dict, str]) -> Union[Dict, None]:
+        """Return a specific document, or `None` if not present."""
+        return self.handle.get(doc)
+
+    def row_count(self) -> int:
+        """Return the number of rows in a table."""
+        return self.handle.count()
+
+    def keys(self) -> Iterable[str]:
+        """Return all the keys in a table."""
+        return self.handle.keys()
+
+    def row_fields(self, filter_keys: bool = False) -> List[str]:
+        """Return a the fields present on each row in this table."""
+        keys = []
+        cur = self.handle.find({}, limit=1)
+
+        # TODO: Determine what should happen if cur is empty
+        if not cur.empty():
+            doc: Dict = next(cur)
+
+            if filter_keys:
+                keys = list(util.filter_unwanted_keys(doc).keys())
+            else:
+                keys = list(doc.keys())
+
+        return keys
+
+    def rename(self, new_name: str) -> None:
+        """Rename a table."""
+        self.handle.rename(new_name)
+        self.name = new_name
+
+    def insert(self, rows: List[Dict]) -> List[Dict]:
+        """
+        Insert rows into this table.
+
+        Returns the metadata from the documents inserted.
+        """
+        return self.handle.insert_many(rows)
+
+    # This may be a reason to split Table up into EdgeTable and NodeTable
+    def edge_properties(self) -> EdgeTableProperties:
+        """
+        Return extracted information about an edge table.
+
+        Extracts 3 pieces of data from an edge table.
+
+        table_keys: A map of all referenced tables to the specific keys referenced.
+        from_tables: A set containing the tables referenced in the _from column.
+        to_tables: A set containing the tables referenced in the _to column.
+
+        Raises an InternalServerError if this table is not an edge table.
+        """
+        props = self.handle.properties()
+        if not props["edge"]:
+            raise InternalServerError(f"Table {self.name} is not an edge table.")
```

Comment:
@JackWilb I see your point regarding it being a 400 error, but there's actually no way to reach this error through the client, except for an error with our classification/handling of the data internally. However, you could probably invoke this intentionally by calling the endpoints directly. Put it in a bit of a weird spot. Specifically, you could invoke this error if you were to call [this](https://api.multinet.app/apidocs/#/graph/post_api_workspaces__workspace__graphs__graph_) endpoint with a node_table in the `edge_table` parameter. In that case, I don't think it should be a `RuntimeError`, since it is the user's fault at that point. This becomes more of an issue when other collaborators are hitting our API directly, instead of just through the client.

For the reasons I listed, I agree with Jack and think it should be a 400 level error. Curious what you think @waxlamp.

Reply:
Good argument for 4xx.

I guess it could be a [422](https://httpstatuses.com/422) but tbh just a 400 with an appropriate message (such as the one you've got here) would be fine.

Reply:
47ec4bd